lms/djangoapps/mobile_api/users/serializers.py

```diff
@@ -141,6 +162,101 @@ class Meta:
         lookup_field = 'username'
 
 
+class CourseEnrollmentSerializerModifiedForPrimary(CourseEnrollmentSerializer):
+    """
+    Serializes CourseEnrollment models for API v4.
+
+    Adds `course_status` field into serializer data.
+    """
+
+    course_status = serializers.SerializerMethodField()
+    course_progress = serializers.SerializerMethodField()
+    course_assignments = serializers.SerializerMethodField()
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.course = modulestore().get_course(self.instance.course.id)
+
+    def get_course_status(self, model: CourseEnrollment) -> Optional[Dict[str, List[str]]]:
+        """
+        Gets course status for the given user's enrollments.
+        """
+        try:
+            block_id = str(get_key_to_last_completed_block(model.user, model.course.id))
+        except UnavailableCompletionData:
+            block_id = ''
+
+        if not block_id:
+            return None
+
+        path, unit_name = self._get_last_visited_block_path_and_unit_name(block_id)
+        if not path and unit_name:
+            return None
+
+        path_ids = [str(block.location) for block in path]
+
+        return {
+            'last_visited_module_id': path_ids[0],
+            'last_visited_module_path': path_ids,
+            'last_visited_block_id': block_id,
+            'last_visited_unit_display_name': unit_name,
+        }
+
+    @staticmethod
+    def _get_last_visited_block_path_and_unit_name(
+        block_id: str
+    ) -> Union[Tuple[None, None], Tuple[List['XBlock'], str]]:  # lint-amnesty, pylint: disable=unused-variable
+        """
+        Returns the path to the latest block and unit name visited by the current user.
+        """
+        try:
+            last_visited_block = modulestore().get_item(UsageKey.from_string(block_id))
+            vertical = last_visited_block.get_parent()
+            sequential = vertical.get_parent()
+            chapter = sequential.get_parent()
+            course = chapter.get_parent()
```

Comment:
Querying blocks from modulstore is very inefficient and should be [avoided whenever possible](https://github.com/openedx/edx-platform/blob/master/docs/decisions/0011-limit-modulestore-use-in-lms.rst). All you need from the vertical, sequential, chapter, and course are their usage keys, right? There must be a more efficient way to get those. Check out the block_structures API.

Reply:
Yes, I rewrote this logic. Now the path to the last visited block is calculated in the same way as on the web. I used `path_to_location` function, this function used for the `Resume` button on web.

Reply:
Looks great, thanks!